logging/src/lib.rs

```diff
@@ -151,108 +150,69 @@ impl<T> Logger<T> {
-    pub fn flush(&mut self) {
+    pub fn flush(&self) {
         <Self as Flush>::flush(self);
     }
 }
 
-impl<T, A: ?Sized + FnMut(&Duration, &mut Vec<(Duration, T)>)> LoggerInner<T, A> {
-
-    /// The upper limit for buffers to allocate, size in bytes. [Self::buffer_capacity] converts
-    /// this to size in elements.
-    const BUFFER_SIZE_BYTES: usize = 1 << 13;
-
-    /// The maximum buffer capacity in elements. Returns a number between [Self::BUFFER_SIZE_BYTES]
-    /// and 1, inclusively.
-    // TODO: This fn is not const because it cannot depend on non-Sized generic parameters
-    fn buffer_capacity() -> usize {
-        let size =  ::std::mem::size_of::<(Duration, T)>();
-        if size == 0 {
-            Self::BUFFER_SIZE_BYTES
-        } else if size <= Self::BUFFER_SIZE_BYTES {
-            Self::BUFFER_SIZE_BYTES / size
-        } else {
-            1
-        }
-    }
-
+impl<CB: ContainerBuilder, A: ?Sized + FnMut(&Duration, &mut CB::Container)> LoggerInner<CB, A> {
     pub fn log_many<I>(&mut self, events: I)
-        where I: IntoIterator, I::Item: Into<T>
+        where I: IntoIterator, CB: PushInto<(Duration, I::Item)>,
     {
         let elapsed = self.time.elapsed() + self.offset;
         for event in events {
-            self.buffer.push((elapsed, event.into()));
-            if self.buffer.len() == self.buffer.capacity() {
-                // Would call `self.flush()`, but for `RefCell` panic.
-                (self.action)(&elapsed, &mut self.buffer);
-                // The buffer clear could plausibly be removed, changing the semantics but allowing users
-                // to do in-place updates without forcing them to take ownership.
-                self.buffer.clear();
-                let buffer_capacity = self.buffer.capacity();
-                if buffer_capacity < Self::buffer_capacity() {
-                    self.buffer.reserve((buffer_capacity+1).next_power_of_two());
-                }
+            self.builder.push_into((elapsed, event.into()));
+            while let Some(container) = self.builder.extract() {
+                (self.action)(&elapsed, container);
             }
         }
     }
+
+    pub fn flush(&mut self) {
+        // TODO: This previously called action with an empty buffer if there was no data.
+        let elapsed = self.time.elapsed() + self.offset;
+        while let Some(buffer) = self.builder.finish() {
+            (self.action)(&elapsed, buffer);
+        }
+    }
 }
 
 /// Flush on the *last* drop of a logger.
-impl<T, A: ?Sized + FnMut(&Duration, &mut Vec<(Duration, T)>)> Drop for LoggerInner<T, A> {
+impl<CB: ContainerBuilder, A: ?Sized + FnMut(&Duration, &mut CB::Container)> Drop for LoggerInner<CB, A> {
     fn drop(&mut self) {
         // Avoid sending out empty buffers just because of drops.
-        if !self.buffer.is_empty() {
-            self.flush();
-        }
+        self.flush();
```

Comment:
The code doesn't really match the comment any more. I think it does avoid sending empty buffers because of `flush()`'s implementation, but also the implementation has a TODO to change that.